# Traffic Sign Classification with Keras

Keras exists to make coding deep neural networks simpler. To demonstrate just how easy it is, you’re going to use Keras to build a convolutional neural network in a few dozen lines of code.

You’ll be connecting the concepts from the previous lessons to the methods that Keras provides.

## Dataset

The network you'll build with Keras is similar to the example in Keras’s GitHub repository that builds out a [convolutional neural network for MNIST](https://github.com/fchollet/keras/blob/master/examples/mnist_cnn.py). 

However, instead of using the [MNIST](http://yann.lecun.com/exdb/mnist/) dataset, you're going to use the [German Traffic Sign Recognition Benchmark](http://benchmark.ini.rub.de/?section=gtsrb&subsection=news) dataset that you've used previously.

You can download pickle files with sanitized traffic sign data here:

In [1]:
from urllib.request import urlretrieve
from os.path import isfile
from tqdm import tqdm

class DLProgress(tqdm):
    last_block = 0

    def hook(self, block_num=1, block_size=1, total_size=None):
        self.total = total_size
        self.update((block_num - self.last_block) * block_size)
        self.last_block = block_num

if not isfile('train.p'):
    with DLProgress(unit='B', unit_scale=True, miniters=1, desc='Train Dataset') as pbar:
        urlretrieve(
            'https://s3.amazonaws.com/udacity-sdc/datasets/german_traffic_sign_benchmark/train.p',
            'train.p',
            pbar.hook)

if not isfile('test.p'):
    with DLProgress(unit='B', unit_scale=True, miniters=1, desc='Test Dataset') as pbar:
        urlretrieve(
            'https://s3.amazonaws.com/udacity-sdc/datasets/german_traffic_sign_benchmark/test.p',
            'test.p',
            pbar.hook)

print('Training and Test data downloaded.')

Train Dataset: 120MB [00:34, 3.49MB/s]                           
Test Dataset: 38.8MB [00:10, 3.86MB/s]                            

Training and Test data downloaded.


## Overview

Here are the steps you'll take to build the network:

1. Load the training data.
2. Preprocess the data.
3. Build a feedforward neural network to classify traffic signs.
4. Build a convolutional neural network to classify traffic signs.
5. Evaluate the final neural network on testing data.

Keep an eye on the network’s accuracy over time. Once the accuracy reaches the 98% range, you can be confident that you’ve built and trained an effective model.

In [3]:
import pickle
import numpy as np
import math

# Fix error with TF and Keras
import tensorflow as tf
tf.python.control_flow_ops = tf

print('Modules loaded.')

Modules loaded.


## Load the Data

Start by importing the data from the pickle file.

In [4]:
with open('train.p', 'rb') as f:
    data = pickle.load(f)

# TODO: Load the feature data to the variable X_train
X_train = data['features']
# TODO: Load the label data to the variable y_train
y_train = data['labels']

In [5]:
# STOP: Do not change the tests below. Your implementation should pass these tests. 
assert np.array_equal(X_train, data['features']), 'X_train not set to data[\'features\'].'
assert np.array_equal(y_train, data['labels']), 'y_train not set to data[\'labels\'].'
print('Tests passed.')

Tests passed.


## Preprocess the Data

1. Shuffle the data
2. Normalize the features using Min-Max scaling between -0.5 and 0.5
3. One-Hot Encode the labels

### Shuffle the data
Hint: You can use the [scikit-learn shuffle](http://scikit-learn.org/stable/modules/generated/sklearn.utils.shuffle.html) function to shuffle the data.

In [6]:
# TODO: Shuffle the data
# Split the data into training  training/validation set


#from sklearn.model_selection import train_test_split
#X_train, X_validation, y_train, y_validation = train_test_split(X_train, y_train,test_size=0.20, random_state=42)


from sklearn.utils import shuffle
X_train, y_train = shuffle(X_train, y_train)

In [7]:
# STOP: Do not change the tests below. Your implementation should pass these tests. 
assert X_train.shape == data['features'].shape, 'X_train has changed shape. The shape shouldn\'t change when shuffling.'
assert y_train.shape == data['labels'].shape, 'y_train has changed shape. The shape shouldn\'t change when shuffling.'
assert not np.array_equal(X_train, data['features']), 'X_train not shuffled.'
assert not np.array_equal(y_train, data['labels']), 'y_train not shuffled.'
print('Tests passed.')

Tests passed.


### Normalize the features
Hint: You solved this in [TensorFlow lab](https://github.com/udacity/CarND-TensorFlow-Lab/blob/master/lab.ipynb) Problem 1.

In [8]:
# TODO: Normalize the data features to the variable X_normalized
center_range = (np.max(X_train)+np.min(X_train))/2.0
range_data = (np.max(X_train)-np.min(X_train))
X_normalized = (np.float32(X_train)-center_range)/(range_data)

In [9]:
# STOP: Do not change the tests below. Your implementation should pass these tests. 
assert math.isclose(np.min(X_normalized), -0.5, abs_tol=1e-5) and math.isclose(np.max(X_normalized), 0.5, abs_tol=1e-5), 'The range of the training data is: {} to {}.  It must be -0.5 to 0.5'.format(np.min(X_normalized), np.max(X_normalized))
print('Tests passed.')

Tests passed.


### One-Hot Encode the labels
Hint: You can use the [scikit-learn LabelBinarizer](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelBinarizer.html) function to one-hot encode the labels.

In [10]:
# TODO: One Hot encode the labels to the variable y_one_hot
from sklearn import preprocessing
lb = preprocessing.LabelBinarizer()
lb.fit(y_train)
print(lb.classes_)
y_one_hot = lb.transform(y_train)
print(y_one_hot.shape)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42]
(39209, 43)


In [11]:
# STOP: Do not change the tests below. Your implementation should pass these tests. 
import collections

assert y_one_hot.shape == (39209, 43), 'y_one_hot is not the correct shape.  It\'s {}, it should be (39209, 43)'.format(y_one_hot.shape)
assert next((False for y in y_one_hot if collections.Counter(y) != {0: 42, 1: 1}), True), 'y_one_hot not one-hot encoded.'
print('Tests passed.')

Tests passed.


## Keras Sequential Model
```python
from keras.models import Sequential

# Create the Sequential model
model = Sequential()
```
The `keras.models.Sequential` class is a wrapper for the neural network model. Just like many of the class models in scikit-learn, it provides common functions like `fit()`, `evaluate()`, and `compile()`.  We'll cover these functions as we get to them.  Let's start looking at the layers of the model.

## Keras Layer
A Keras layer is just like a neural network layer.  It can be fully connected, max pool, activation, etc.  You can add a layer to the model using the model's `add()` function.  For example, a simple model would look like this:
```python
from keras.models import Sequential
from keras.layers.core import Dense, Activation, Flatten

# Create the Sequential model
model = Sequential()

# 1st Layer - Add a flatten layer
model.add(Flatten(input_shape=(32, 32, 3)))

# 2nd Layer - Add a fully connected layer
model.add(Dense(100))

# 3rd Layer - Add a ReLU activation layer
model.add(Activation('relu'))

# 4th Layer - Add a fully connected layer
model.add(Dense(60))

# 5th Layer - Add a ReLU activation layer
model.add(Activation('relu'))
```
Keras will automatically infer the shape of all layers after the first layer.  This means you only have to set the input dimensions for the first layer.

The first layer from above, `model.add(Flatten(input_shape=(32, 32, 3)))`, sets the input dimension to (32, 32, 3) and output dimension to (3072=32\*32\*3).  The second layer takes in the output of the first layer and sets the output dimenions to (100).  This chain of passing output to the next layer continues until the last layer, which is the output of the model.

## Build a Multi-Layer Feedforward Network

Build a multi-layer feedforward neural network to classify the traffic sign images.

1. Set the first layer to a `Flatten` layer with the `input_shape` set to (32, 32, 3)
2. Set the second layer to `Dense` layer width to 128 output. 
3. Use a ReLU activation function after the second layer.
4. Set the output layer width to 43, since there are 43 classes in the dataset.
5. Use a softmax activation function after the output layer.

To get started, review the Keras documentation about [models](https://keras.io/models/sequential/) and [layers](https://keras.io/layers/core/).

The Keras example of a [Multi-Layer Perceptron](https://github.com/fchollet/keras/blob/master/examples/mnist_mlp.py) network is similar to what you need to do here. Use that as a guide, but keep in mind that there are a number of differences.

In [12]:
from keras.models import Sequential
from keras.layers.core import Dense, Activation, Flatten
from keras.activations import relu, softmax

model = Sequential()
# TODO: Build a Multi-layer feedforward neural network with Keras here.
model.add(Flatten(input_shape=(32, 32, 3)))
model.add(Dense(128))
model.add(Activation('relu'))
model.add(Dense(43))
model.add(Activation('softmax'))


In [33]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
from keras.layers.core import Dense, Activation, Flatten
from keras.activations import relu, softmax

def check_layers(layers, true_layers):
    assert len(true_layers) != 0, 'No layers found'
    for layer_i in range(len(layers)):
        assert isinstance(true_layers[layer_i], layers[layer_i]), 'Layer {} is not a {} layer'.format(layer_i+1, layers[layer_i].__name__)
    assert len(true_layers) == len(layers), '{} layers found, should be {} layers'.format(len(true_layers), len(layers))

check_layers([Flatten, Dense, Activation, Dense, Activation], model.layers)

assert model.layers[0].input_shape == (None, 32, 32, 3), 'First layer input shape is wrong, it should be (32, 32, 3)'
assert model.layers[1].output_shape == (None, 128), 'Second layer output is wrong, it should be (128)'
assert model.layers[2].activation == relu, 'Third layer not a relu activation layer'
assert model.layers[3].output_shape == (None, 43), 'Fourth layer output is wrong, it should be (43)'
assert model.layers[4].activation == softmax, 'Fifth layer not a softmax activation layer'
print('Tests passed.')

Tests passed.


## Training a Sequential Model
You built a multi-layer neural network in Keras, now let's look at training a neural network.
```python
from keras.models import Sequential
from keras.layers.core import Dense, Activation

model = Sequential()
...

# Configures the learning process and metrics
model.compile('sgd', 'mean_squared_error', ['accuracy'])

# Train the model
# History is a record of training loss and metrics
history = model.fit(x_train_data, Y_train_data, batch_size=128, nb_epoch=2, validation_split=0.2)

# Calculate test score
test_score = model.evaluate(x_test_data, Y_test_data)
```
The code above configures, trains, and tests the model.  The line `model.compile('sgd', 'mean_squared_error', ['accuracy'])` configures the model's optimizer to `'sgd'`(stochastic gradient descent), the loss to `'mean_squared_error'`, and the metric to `'accuracy'`.  

You can find more optimizers [here](https://keras.io/optimizers/), loss functions [here](https://keras.io/objectives/#available-objectives), and more metrics [here](https://keras.io/metrics/#available-metrics).

To train the model, use the `fit()` function as shown in `model.fit(x_train_data, Y_train_data, batch_size=128, nb_epoch=2, validation_split=0.2)`.  The `validation_split` parameter will split a percentage of the training dataset to be used to validate the model.  The model can be further tested with the test dataset using the `evaluation()` function as shown in the last line.

## Train the Network

1. Compile the network using adam optimizer and categorical_crossentropy loss function.
2. Train the network for ten epochs and validate with 20% of the training data.

In [43]:
# TODO: Compile and train the model here.
# Configures the learning process and metrics
model.compile('adam', 'categorical_crossentropy', ['accuracy'])

# Train the model
# History is a record of training loss and metrics
history = model.fit(X_normalized, y_one_hot, batch_size=128, nb_epoch=10, validation_split=0.2)

# Calculate test score
test_score = model.evaluate(X_normalized, y_one_hot)


Train on 31367 samples, validate on 7842 samples
Epoch 1/10
31367/31367 [==============================] - 1s - loss: 0.6256 - acc: 0.8382 - val_loss: 0.6590 - val_acc: 0.7869
Epoch 2/10
31367/31367 [==============================] - 1s - loss: 0.4852 - acc: 0.8748 - val_loss: 0.5298 - val_acc: 0.8358
Epoch 3/10
31367/31367 [==============================] - 1s - loss: 0.4071 - acc: 0.8957 - val_loss: 0.4358 - val_acc: 0.8906
Epoch 4/10
31367/31367 [==============================] - 1s - loss: 0.3471 - acc: 0.9107 - val_loss: 0.3418 - val_acc: 0.9180
Epoch 5/10
31367/31367 [==============================] - 0s - loss: 0.3283 - acc: 0.9138 - val_loss: 0.4586 - val_acc: 0.8583
Epoch 6/10
31367/31367 [==============================] - 1s - loss: 0.2794 - acc: 0.9265 - val_loss: 0.4095 - val_acc: 0.8698
Epoch 7/10
31367/31367 [==============================] - 0s - loss: 0.2726 - acc: 0.9272 - val_loss: 0.3865 - val_acc: 0.8914
Epoch 8/10
31367/31367 [==============================] - 0s -

In [44]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
from keras.optimizers import Adam

assert model.loss == 'categorical_crossentropy', 'Not using categorical_crossentropy loss function'
assert isinstance(model.optimizer, Adam), 'Not using adam optimizer'
assert len(history.history['acc']) == 10, 'You\'re using {} epochs when you need to use 10 epochs.'.format(len(history.history['acc']))

assert history.history['acc'][-1] > 0.92, 'The training accuracy was: %.3f. It shoud be greater than 0.92' % history.history['acc'][-1]
assert history.history['val_acc'][-1] > 0.85, 'The validation accuracy is: %.3f. It shoud be greater than 0.85' % history.history['val_acc'][-1]
print('Tests passed.')

Tests passed.


## Convolutions
1. Re-construct the previous network
2. Add a [convolutional layer](https://keras.io/layers/convolutional/#convolution2d) with 32 filters, a 3x3 kernel, and valid padding before the flatten layer.
3. Add a ReLU activation after the convolutional layer.

Hint 1: The Keras example of a [convolutional neural network](https://github.com/fchollet/keras/blob/master/examples/mnist_cnn.py) for MNIST would be a good example to review.

In [51]:
# TODO: Re-construct the network and add a convolutional layer before the flatten layer.
from keras.models import Sequential
from keras.layers.core import Dense, Activation, Flatten
from keras.activations import relu, softmax
from keras.layers.convolutional import Convolution2D

model = Sequential()
model.add(Convolution2D(32, 3, 3, border_mode='valid', input_shape=(32, 32, 3)))
model.add(Activation('relu'))
model.add(Flatten())
model.add(Dense(128))
model.add(Activation('relu'))
model.add(Dense(43))
model.add(Activation('softmax'))

In [53]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
from keras.layers.core import Dense, Activation, Flatten
from keras.layers.convolutional import Convolution2D

check_layers([Convolution2D, Activation, Flatten, Dense, Activation, Dense, Activation], model.layers)

assert model.layers[0].input_shape == (None, 32, 32, 3), 'First layer input shape is wrong, it should be (32, 32, 3)'
assert model.layers[0].nb_filter == 32, 'Wrong number of filters, it should be 32'
assert model.layers[0].nb_col == model.layers[0].nb_row == 3, 'Kernel size is wrong, it should be a 3x3'
assert model.layers[0].border_mode == 'valid', 'Wrong padding, it should be valid'

model.compile('adam', 'categorical_crossentropy', ['accuracy'])
history = model.fit(X_normalized, y_one_hot, batch_size=128, nb_epoch=2, validation_split=0.2)
assert(history.history['val_acc'][-1] > 0.91), "The validation accuracy is: %.3f.  It should be greater than 0.91" % history.history['val_acc'][-1]
print('Tests passed.')

Train on 31367 samples, validate on 7842 samples
Epoch 1/10
31367/31367 [==============================] - 2s - loss: 0.1720 - acc: 0.9573 - val_loss: 0.1781 - val_acc: 0.9510
Epoch 2/10
31367/31367 [==============================] - 2s - loss: 0.0907 - acc: 0.9784 - val_loss: 0.1538 - val_acc: 0.9566
Epoch 3/10
31367/31367 [==============================] - 2s - loss: 0.0671 - acc: 0.9833 - val_loss: 0.1239 - val_acc: 0.9685
Epoch 4/10
31367/31367 [==============================] - 2s - loss: 0.0507 - acc: 0.9878 - val_loss: 0.1334 - val_acc: 0.9690
Epoch 5/10
31367/31367 [==============================] - 2s - loss: 0.0374 - acc: 0.9905 - val_loss: 0.1096 - val_acc: 0.9737
Epoch 6/10
31367/31367 [==============================] - 2s - loss: 0.0283 - acc: 0.9936 - val_loss: 0.1458 - val_acc: 0.9663
Epoch 7/10
31367/31367 [==============================] - 2s - loss: 0.0372 - acc: 0.9899 - val_loss: 0.1270 - val_acc: 0.9703
Epoch 8/10
31367/31367 [==============================] - 2s -

## Pooling
1. Re-construct the network
2. Add a 2x2 [max pooling layer](https://keras.io/layers/pooling/#maxpooling2d) immediately following your convolutional layer.

In [56]:
# TODO: Re-construct the network and add a pooling layer after the convolutional layer.
from keras.models import Sequential
from keras.layers.core import Dense, Activation, Flatten
from keras.activations import relu, softmax
from keras.layers.convolutional import Convolution2D
from keras.layers.pooling import MaxPooling2D

model = Sequential()
model.add(Convolution2D(32, 3, 3, border_mode='valid', input_shape=(32, 32, 3)))
model.add(MaxPooling2D(pool_size=(2, 2), strides=None, border_mode='same'))
model.add(Activation('relu'))
model.add(Flatten())
model.add(Dense(128))
model.add(Activation('relu'))
model.add(Dense(43))
model.add(Activation('softmax'))


In [57]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
from keras.layers.core import Dense, Activation, Flatten
from keras.layers.convolutional import Convolution2D
from keras.layers.pooling import MaxPooling2D

check_layers([Convolution2D, MaxPooling2D, Activation, Flatten, Dense, Activation, Dense, Activation], model.layers)
assert model.layers[1].pool_size == (2, 2), 'Second layer must be a max pool layer with pool size of 2x2'

model.compile('adam', 'categorical_crossentropy', ['accuracy'])
history = model.fit(X_normalized, y_one_hot, batch_size=128, nb_epoch=2, validation_split=0.2)
assert(history.history['val_acc'][-1] > 0.91), "The validation accuracy is: %.3f.  It should be greater than 0.91" % history.history['val_acc'][-1]
print('Tests passed.')

Train on 31367 samples, validate on 7842 samples
Epoch 1/2
31367/31367 [==============================] - 2s - loss: 1.5232 - acc: 0.6022 - val_loss: 0.7684 - val_acc: 0.7724
Epoch 2/2
31367/31367 [==============================] - 1s - loss: 0.4735 - acc: 0.8836 - val_loss: 0.3184 - val_acc: 0.9300
Tests passed.


## Dropout
1. Re-construct the network
2. Add a [dropout](https://keras.io/layers/core/#dropout) layer after the pooling layer. Set the dropout rate to 50%.

In [59]:
# TODO: Re-construct the network and add dropout after the pooling layer.
from keras.models import Sequential
from keras.layers.core import Dense, Activation, Flatten, Dropout
from keras.activations import relu, softmax
from keras.layers.convolutional import Convolution2D
from keras.layers.pooling import MaxPooling2D

model = Sequential()
model.add(Convolution2D(32, 3, 3, border_mode='valid', input_shape=(32, 32, 3)))
model.add(MaxPooling2D(pool_size=(2, 2), strides=None, border_mode='same'))
model.add(Dropout(0.5))
model.add(Activation('relu'))
model.add(Flatten())
model.add(Dense(128))
model.add(Activation('relu'))
model.add(Dense(43))
model.add(Activation('softmax'))


In [60]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
from keras.layers.core import Dense, Activation, Flatten, Dropout
from keras.layers.convolutional import Convolution2D
from keras.layers.pooling import MaxPooling2D

check_layers([Convolution2D, MaxPooling2D, Dropout, Activation, Flatten, Dense, Activation, Dense, Activation], model.layers)
assert model.layers[2].p == 0.5, 'Third layer should be a Dropout of 50%'

model.compile('adam', 'categorical_crossentropy', ['accuracy'])
history = model.fit(X_normalized, y_one_hot, batch_size=128, nb_epoch=2, validation_split=0.2)
assert(history.history['val_acc'][-1] > 0.91), "The validation accuracy is: %.3f.  It should be greater than 0.91" % history.history['val_acc'][-1]
print('Tests passed.')

Train on 31367 samples, validate on 7842 samples
Epoch 1/2
31367/31367 [==============================] - 2s - loss: 1.5548 - acc: 0.5855 - val_loss: 0.6292 - val_acc: 0.8476
Epoch 2/2
31367/31367 [==============================] - 2s - loss: 0.4970 - acc: 0.8717 - val_loss: 0.3270 - val_acc: 0.9323
Tests passed.


## Optimization
Congratulations! You've built a neural network with convolutions, pooling, dropout, and fully-connected layers, all in just a few lines of code.

Have fun with the model and see how well you can do! Add more layers, or regularization, or different padding, or batches, or more training epochs.

What is the best validation accuracy you can achieve?

In [ ]:
# TODO: Build a model
from keras.models import Sequential, Model
from keras.layers.core import Dense, Activation, Flatten, Dropout
from keras.layers import Input
from keras.activations import relu, softmax
from keras.layers.convolutional import Convolution2D
from keras.layers.pooling import MaxPooling2D
from keras.layers.normalization import BatchNormalization
from keras.engine.topology import Merge

model_input = Input(shape=(32, 32, 3))
model_stem = Convolution2D(32, 3, 3, border_mode='valid', input_shape=(32, 32, 3))(model_input)
model_stem = MaxPooling2D(pool_size=(2, 2), strides=None, border_mode='same') (model_stem)
model_stem = Dropout(0.5)(model_stem)
branch_input = Activation('elu')(model_stem)

left_branch = Convolution2D(32, 1 , 1,border_mode='same')(branch_input)

right_branch = Convolution2D(32, 7, 1, border_mode='same')  (branch_input)
right_branch = BatchNormalization(epsilon=0.001, mode=0, axis=3, momentum=0.9) (right_branch)
right_branch = Activation('elu') (right_branch)
right_branch = Convolution2D(32, 1, 7, border_mode='same') (right_branch)
right_branch = BatchNormalization(epsilon=0.001, mode=0, axis=3, momentum=0.9) (right_branch)
right_branch = Activation('elu') (right_branch)

#merged = merge([left_branch, right_branch], mode='concat', concat_axis=3) 
merged = merge([left_branch, right_branch], mode='sum') 

merged = Convolution2D(64, 3, 3, border_mode='valid', activation='elu')(merged)
merged = BatchNormalization(epsilon=0.001, mode=0, axis=-1, momentum=0.9)(merged)
merged = Convolution2D(64, 3, 3, border_mode='valid', activation='elu')(merged)
merged = Flatten()(merged)
merged = Dense(128, activation='elu')(merged)
merged = Dense(43, activation='softmax')(merged)

# TODO: Compile and train the model
model =  Model(input=model_input, output=merged) 
model.compile('adam', 'categorical_crossentropy', ['accuracy'])

# Train the model
# History is a record of training loss and metrics
history = model.fit(X_normalized, y_one_hot, batch_size=256, nb_epoch=30, validation_split=0.2)

# Calculate test score
test_score = model.evaluate(X_normalized, y_one_hot)

Train on 31367 samples, validate on 7842 samples
Epoch 1/30
31367/31367 [==============================] - 5s - loss: 0.7936 - acc: 0.7755 - val_loss: 0.1899 - val_acc: 0.9468
Epoch 2/30
31367/31367 [==============================] - 4s - loss: 0.1620 - acc: 0.9505 - val_loss: 0.1019 - val_acc: 0.9723
Epoch 3/30
31367/31367 [==============================] - 4s - loss: 0.0867 - acc: 0.9728 - val_loss: 0.0908 - val_acc: 0.9763
Epoch 4/30
31367/31367 [==============================] - 4s - loss: 0.0601 - acc: 0.9817 - val_loss: 0.0868 - val_acc: 0.9791
Epoch 5/30
31367/31367 [==============================] - 4s - loss: 0.0544 - acc: 0.9827 - val_loss: 0.0630 - val_acc: 0.9852
Epoch 6/30
31367/31367 [==============================] - 4s - loss: 0.0448 - acc: 0.9859 - val_loss: 0.0613 - val_acc: 0.9848
Epoch 7/30
31367/31367 [==============================] - 4s - loss: 0.0421 - acc: 0.9857 - val_loss: 0.0675 - val_acc: 0.9839
Epoch 8/30
31367/31367 [==============================] - 4s -

In [66]:
for layer in model.layers:
    print('Layer: {}\tInput shape {}\tOutput shape {}'.format(layer.name, layer.input_shape, layer.output_shape) )

print(model.layers[-1].input_shape[1:])

from keras.utils.visualize_util import plot
plot(model, to_file='model.png', show_shapes=True)

Layer: input_27	Input shape (None, 32, 32, 3)	Output shape (None, 32, 32, 3)
Layer: convolution2d_104	Input shape (None, 32, 32, 3)	Output shape (None, 30, 30, 32)
Layer: maxpooling2d_31	Input shape (None, 30, 30, 32)	Output shape (None, 15, 15, 32)
Layer: dropout_31	Input shape (None, 15, 15, 32)	Output shape (None, 15, 15, 32)
Layer: activation_71	Input shape (None, 15, 15, 32)	Output shape (None, 15, 15, 32)
Layer: convolution2d_106	Input shape (None, 15, 15, 32)	Output shape (None, 15, 15, 32)
Layer: batchnormalization_41	Input shape (None, 15, 15, 32)	Output shape (None, 15, 15, 32)
Layer: activation_72	Input shape (None, 15, 15, 32)	Output shape (None, 15, 15, 32)
Layer: convolution2d_107	Input shape (None, 15, 15, 32)	Output shape (None, 15, 15, 32)
Layer: batchnormalization_42	Input shape (None, 15, 15, 32)	Output shape (None, 15, 15, 32)
Layer: convolution2d_105	Input shape (None, 15, 15, 32)	Output shape (None, 15, 15, 32)
Layer: activation_73	Input shape (None, 15, 15, 32)	O

<img src="model.png" alt="Model"/>

**Best Validation Accuracy:** (99.1%)

## Testing
Once you've picked out your best model, it's time to test it.

Load up the test data and use the [`evaluate()` method](https://keras.io/models/model/#evaluate) to see how well it does.

Hint 1: The `evaluate()` method should return an array of numbers. Use the [`metrics_names`](https://keras.io/models/model/) property to get the labels.

In [67]:
# TODO: Load test data
with open('test.p', 'rb') as f:
    data = pickle.load(f)

X_test = data['features']
y_test = data['labels']
    
# TODO: Preprocess data & one-hot encode the labels
# TODO: Normalize the data features to the variable X_normalized
X_test_normalized = (np.float32(X_test)-center_range)/(range_data)

# TODO: One Hot encode the labels to the variable y_one_hot
from sklearn import preprocessing
lb = preprocessing.LabelBinarizer()
lb.fit(y_test)
print(lb.classes_)
y_test_one_hot = lb.transform(y_test)
print(y_test_one_hot.shape)


[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42]
(12630, 43)
12608/12630 [============================>.] - ETA: 0s

[0.42564863672882602, 0.93911322258052987]

In [72]:
# Evaluate method
model.evaluate(X_test_normalized, y_test_one_hot, batch_size=32, verbose=1, sample_weight=None)

12544/12630 [============================>.] - ETA: 0s

[0.36836492733280413, 0.95241488526005269]

In [68]:
print(model.metrics_names)


['loss', 'acc']


**Test Accuracy:** (95.2%)

## Summary
Keras is a great tool to use if you want to quickly build a neural network and evaluate performance.